src/groupeddataframe/utils.jl

```diff
@@ -323,54 +326,103 @@ function row_group_slots(cols::NTuple{N, AbstractVector},
-                j = 0
-            else
-                seen[j] = true
+        @spawn_for_chunks 1_000_000 for i in eachindex(groups)
+            @inbounds begin
+                local refs_i
+                let i=i # Workaround for julia#15276
+                    refs_i = map(c -> c[i], refarrays)
+                end
+                vals = map((m, r, s, fi) -> m[r-fi+1] * s, refmaps, refs_i, strides, firstinds)
+                j = sum(vals) + 1
+                # x < 0 happens with -1 in refmap, which corresponds to missing
+                if skipmissing && any(x -> x < 0, vals)
+                    j = 0
+                end
+                groups[i] = j
             end
-            groups[i] = j
         end
     else
-        @inbounds @spawn_for_chunks 1_000_000 for i in eachindex(groups)
-            local refs_i
-            let i=i # Workaround for julia#15276
-                refs_i = map(refarrays, missinginds) do ref, missingind
-                    r = Int(ref[i])
-                    if skipmissing
-                        return r == missingind ? -1 : (r > missingind ? r-1 : r)
-                    else
-                        return r
+        @spawn_for_chunks 1_000_000 for i in eachindex(groups)
+            @inbounds begin
+                local refs_i
+                let i=i # Workaround for julia#15276
+                    refs_i = map(refarrays, missinginds) do ref, missingind
+                        r = Int(ref[i])
+                        if skipmissing
+                            return r == missingind ? -1 : (r > missingind ? r-1 : r)
+                        else
+                            return r
+                        end
                     end
                 end
+                vals = map((r, s, fi) -> (r-fi) * s, refs_i, strides, firstinds)
+                j = sum(vals) + 1
+                # x < 0 happens with -1, which corresponds to missing
+                if skipmissing && any(x -> x < 0, vals)
+                    j = 0
+                end
+                groups[i] = j
+            end
+        end
+    end
+
+    seen1 = fill(false, ngroups)
+
+    # do division to avoid overflow
+    # we compare the amount of work do be done by both algorithms
+    # assuming less than linear scaling by a factor of 2
+    lg = length(groups)
+    nt = Threads.nthreads()
+    if nt == 1 || lg <= 1_000_000 || ngroups > lg * (1.0 - 2.0 / nt) / nt
+        if skipmissing
+            for j in groups
+                if j > 0
+                    @inbounds seen1[j] = true
+                end
+            end
+        else
+            for j in groups
+                @inbounds seen1[j] = true
+            end
+        end
+    else
+        seen = Vector{Vector{Bool}}(undef, nt)
+        seen[1] = seen1
+        for i in 2:nt
+            seen[i] = fill(false, ngroups)
+        end
+
+        if skipmissing
+            @spawn_for_chunks 1_000_000 for i in eachindex(groups)
```

Comment:
Why not merge this with the loop computing `groups` above? That should reduce the overhead of spawning tasks, and make better use of CPU.

Reply:
This part is only executed if `nt == 1 || lg <= 1_000_000 || ngroups > lg * (1.0 - 2.0 / nt) / nt` is `false`. I could integrate it into the loop above, but this would cause a lot of code duplication (as we would need two versions of the loop above: one with integrated `seen` and other with `seen` extracted out).

Do you think it is worth it?

Reply:
Some benchmarking shows it is worth it, so I will split the code tomorrow.

Reply:
Maybe there's a way to avoid code duplication? Indeed for testing it would be better to share as much code as possible, especially given how complex is the condition. Maybe if you check whether `seeni` is `nothing` inside the loop, LLVM will hoist the check so that performance isn't affected?

Reply:
yes - I am working on this. These things are tricky as you know :).